### PR TITLE
[SYCLomatic] Fix the issue during the migration of c source file

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -11619,7 +11624,19 @@ void CMemoryAPIRule::runRule(const MatchFinder::MatchResult &Result) {
   auto ICE = getNodeAsType<ImplicitCastExpr>(Result, "implicitCast");
   if (!ICE)
     return;
-
+  auto FilePath = DpctGlobalInfo::getLocInfo(ICE->getBeginLoc()).first;
+  auto Extension = path::extension(FilePath.getPath());
+  if (Extension == ".c") {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp]: should we also check if "-x lang" option used? 
also .C

**Review by ziranzha — APPROVED** [review]

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 2]
```diff
@@ -4551,6 +4553,10 @@ std::string DpctGlobalInfo::getStringForRegexReplacement(StringRef MatchedStr) {
   case 'Q':
     return getStringForRegexDefaultQueueAndDevice(
         HelperFuncType::HFT_DefaultQueue, Index);
+  case 'E': {
+    auto &Vec = DpctGlobalInfo::getInstance().getCSourceFileInfo();
+    return Vec[Index]->hasCUDASyntax() ? "c.dp.cpp" : "c";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AnalysisInfo.cpp]: now dp.cpp is not a fix file ext,  you need call the API to get the postfix, it maybe "dp.cpp, cpp, sycl.cpp". 
see more detail in the PR: https://github.com/oneapi-src/SYCLomatic/pull/1532

File: clang/lib/DPCT/SaveNewFiles.cpp [hunk 3]
```diff
@@ -156,16 +156,26 @@ void rewriteFileName(clang::tooling::UnifiedPath &FileName,
...
+    if (FileType & SPT_CudaSource) {
       path::replace_extension(CanonicalPathStr,
                               DpctGlobalInfo::getSYCLSourceExtension());
-    else if (FileType & SPT_CppSource)
-      path::replace_extension(CanonicalPathStr,
-                              Extension +
-                                  DpctGlobalInfo::getSYCLSourceExtension());
-    else if (FileType & SPT_CudaHeader)
+    } else if (FileType & SPT_CppSource) {
+      if (Extension == ".c") {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/SaveNewFiles.cpp:172]: Please also consider  ".C" or a function:  is_c_code()

File: clang/lib/DPCT/Utility.cpp [hunk 4]
```diff
@@ -267,6 +267,9 @@ SourceProcessType GetSourceFileType(const clang::tooling::UnifiedPath &SourcePat
   if (Extension == ".cuh") {
     return SPT_CudaHeader;
   }
+  if (Extension == ".c") {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Utility.cpp:270]: same as above.

File: clang/lib/DPCT/SaveNewFiles.cpp (hunk 3 above)
Comment (intwanghao) [reply clang/lib/DPCT/SaveNewFiles.cpp:172]: The .C(capital letter) usually means C++ source file extension.
https://gcc.gnu.org/onlinedocs/gcc/Invoking-G_002b_002b.html

File: clang/lib/DPCT/Utility.cpp (hunk 4 above)
Comment (intwanghao) [reply clang/lib/DPCT/Utility.cpp:270]: The .C(capital letter) usually means C++ source file extension.
https://gcc.gnu.org/onlinedocs/gcc/Invoking-G_002b_002b.html

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**